src/subdivide_multi.h

```diff
@@ -0,0 +1,302 @@
+///
+///
+bool subTet(const std::array<std::array<double, 3>,4> &pts,
+            const llvm_vecsmall::SmallVector<std::array<double,4>, 20> &vals,
+            const llvm_vecsmall::SmallVector<std::array<std::array<double, 3>,4>, 20> &grads,
+            const double threshold,
+            bool& active);
+/// Two checks for each grid element: zero-crossing check and distance check. This function performs these two checks under the setting of material interface(MI) and its curve networks.
+///
+/// @param[in] pts          4 arrays of three-tuples represent the coordinate of tet vertices. Each subarray represents a coordinate.
+/// @param[in] vals         functions' values at each vertex. Each subarray represents one function's values at four vertices.
+/// @param[in] grads            functions' gradients at each vertices. Each subarray represents one functions' gradients at four vertices and in each axis-aligned partial derivative.
+/// @param[in] threshold            The user-input epsilon value for the distance check.
+/// @param[out] active          A `bool` represents whether the tet is containing part of the geometry, i.e., this tet passes the zero-crossing test.
+///
+/// @return         A `bool` represents whether the tet is "refinable", i.e., passing the zero-crossing test and contains error greater than `threshold`.
+///
+///
+bool subMI(const std::array<std::array<double, 3>,4> &pts,
+           const llvm_vecsmall::SmallVector<std::array<double,4>, 20> &vals,
+           const llvm_vecsmall::SmallVector<std::array<std::array<double, 3>,4>, 20> &grads,
+           const double threshold,
+           bool& active);
+
+/// Enums for the current settings of implicit complexes
+enum geo_obj {
+    IA,
+    CSG,
+    MI
+};
+
+/// Enums for CSG boolean operations
+enum csg_operations{
+    Intersection,
+    Union,
+    Negation
+};
+
+
+///Defines the enumeration of CSG operations. In the data structure, Intersection is 0, Union is 1, and Negation is 2.
+struct csg_unit{
+    int operation;
+    std::array<int, 2> elements;
+};
+
+///load the csg file
+///
+///@param[in] filename          The name of the input CSG tree
+///@param[out] tree         The loaded tree structure
+///
+bool load_csgTree(const std::string filename, llvm_vecsmall::SmallVector<csg_unit, 20>& tree);
+
+/// an iterative algortihm that traverses through the csg tree
+///
+///@param[in] csgTree           The CSG structure: a list of csg units
+///@param[in] curNode           The current index in the csg structure
+///@param[in] funcInt           The intervals of all the functions
+///
+///@param[out] std::pair
+std::pair<std::array<double, 2>, llvm_vecsmall::SmallVector<int, 20>> iterTree(const llvm_vecsmall::SmallVector<csg_unit, 20>csgTree,const int curNode,const llvm_vecsmall::SmallVector<std::array<double , 2>, 20> funcInt);
```

Comment:
What are the returned values of this function? The documentation is not very clear.